plugins/destination/postgresql/client/migrate.go

```diff
@@ -48,165 +130,98 @@ func (c *Client) Migrate(ctx context.Context, tables schema.Tables) error {
 	}
 
-	if reCreatePrimaryKeys {
-		c.logger.Info().Str("table", table.Name).Msg("Recreating primary keys")
-		if err := c.setNotNullOnPks(ctx, table); err != nil {
-			return fmt.Errorf("failed to enforce not null on primary keys: %w", err)
+	sql := "alter table " + tableName + " add constraint primary key (" + strings.Join(table.PrimaryKeys(), ",") + ")"
+	if _, err := tx.Exec(ctx, sql); err != nil {
+		if err := tx.Rollback(ctx); err != nil {
+			c.logger.Error().Err(err).Msg("failed to rollback transaction")
 		}
+		return fmt.Errorf("failed to add primary key constraint on table %s: %w", table.Name, err)
+	}
+	if err := tx.Commit(ctx); err != nil {
+		return fmt.Errorf("failed to commit transaction to recreate primary keys: %w", err)
+	}
+	return nil
+}
 
-		tx, err := c.conn.Begin(ctx)
-		if err != nil {
-			return fmt.Errorf("failed to begin transaction to recreate primary keys: %w", err)
+func (c *Client) autoMigrateTable(ctx context.Context, pgTable *schema.Table, table *schema.Table) error {
+	if changedColumns := table.GetChangedColumns(pgTable); changedColumns != nil {
+		if c.spec.MigrateMode != specs.MigrateModeForced {
+			return fmt.Errorf("postgres table %s has changed %v columns, use --force to drop the column", table.Name, changedColumns)
 		}
-		if _, err := tx.Exec(ctx, dropConstraintSQL); err != nil {
-			if err := tx.Rollback(ctx); err != nil {
-				c.logger.Error().Err(err).Msg("failed to rollback transaction")
+		for _, col := range changedColumns {
+			if err := c.alterColumn(ctx, table.Name, col); err != nil {
+				return err
 			}
-			return fmt.Errorf("failed to drop primary key constraint on table %s: %w", table.Name, err)
 		}
+	}
 
-		sql := "alter table " + tableName + " add constraint " + constraintName + " primary key (" + strings.Join(table.PrimaryKeys(), ",") + ")"
-		if _, err := tx.Exec(ctx, sql); err != nil {
-			if err := tx.Rollback(ctx); err != nil {
-				c.logger.Error().Err(err).Msg("failed to rollback transaction")
+	if addedColumns := table.GetAddedColumns(pgTable); addedColumns != nil {
+		for _, col := range addedColumns {
+			if err := c.addColumn(ctx, table.Name, col); err != nil {
+				return err
 			}
-			return fmt.Errorf("failed to add primary key constraint on table %s: %w", table.Name, err)
-		}
-		if err := tx.Commit(ctx); err != nil {
-			return fmt.Errorf("failed to commit transaction to recreate primary keys: %w", err)
 		}
 	}
 
-	return nil
-}
-
-func (c *Client) setNotNullOnPks(ctx context.Context, table *schema.Table) error {
-	for _, col := range table.PrimaryKeys() {
-		sql := "alter table " + pgx.Identifier{table.Name}.Sanitize() + " alter column " + pgx.Identifier{col}.Sanitize() + " set not null"
-		if _, err := c.conn.Exec(ctx, sql); err != nil {
-			return fmt.Errorf("failed to set not null on column %s on table %s: %w", col, table.Name, err)
+	if c.enabledPks() && !table.IsPrimaryKeyEqual(pgTable) {
```

Comment:
I think we may want to move this check to be above the `GetAddedColumns` check, so that no changes are made to the table if `migrate-forced` is false and a forced migration is necessary. Otherwise the table can be left in a half-migrated state

Reply:
why it's half baked? but I can't move it up because potentially the primary keys are on the columns that were just added.

Reply:
ok yeah you are right. I now moved that check up to fail early.